src/lib/optimizer/join_ordering/join_graph_builder.cpp

```diff
@@ -0,0 +1,168 @@
+        _predicates.emplace_back(std::make_shared<JoinPlanAtomicPredicate>(
+            predicate_node->column_reference(), predicate_node->predicate_condition(), predicate_node->value()));
+      }
+
+      _traverse(node->left_child());
+    } break;
+
+    case LQPNodeType::Union: {
+      /**
+       * A UnionNode is the entry point to disjunction, which is parsed starting from _parse_union(). Normal traversal
+       * is commenced from the node "below" the Union.
+       */
+
+      const auto union_node = std::static_pointer_cast<UnionNode>(node);
+
+      if (union_node->union_mode() == UnionMode::Positions) {
+        const auto parse_result = _parse_union(union_node);
+
+        _traverse(parse_result.base_node);
+        _predicates.emplace_back(parse_result.predicate);
+      } else {
+        _vertices.emplace_back(node);
+      }
+    } break;
+
+    default: { Fail("Node type not suited for JoinGraph"); }
+  }
+}
+
+JoinGraphBuilder::PredicateParseResult JoinGraphBuilder::_parse_predicate(
+    const std::shared_ptr<AbstractLQPNode>& node) const {
+  if (node->type() == LQPNodeType::Predicate) {
+    const auto predicate_node = std::static_pointer_cast<const PredicateNode>(node);
+
+    std::shared_ptr<const AbstractJoinPlanPredicate> left_predicate;
+
+    if (predicate_node->value2()) {
+      DebugAssert(predicate_node->predicate_condition() == PredicateCondition::Between, "Expected between");
+
+      left_predicate = std::make_shared<JoinPlanLogicalPredicate>(
+          std::make_shared<JoinPlanAtomicPredicate>(predicate_node->column_reference(),
+                                                    PredicateCondition::GreaterThanEquals, predicate_node->value()),
+          JoinPlanPredicateLogicalOperator::And,
+          std::make_shared<JoinPlanAtomicPredicate>(predicate_node->column_reference(),
+                                                    PredicateCondition::LessThanEquals, *predicate_node->value2()));
+    } else {
+      left_predicate = std::make_shared<JoinPlanAtomicPredicate>(
+          predicate_node->column_reference(), predicate_node->predicate_condition(), predicate_node->value());
+    }
+
+    const auto base_node = predicate_node->left_child();
+
+    if (base_node->parent_count() > 1) {
+      return {base_node, left_predicate};
+    } else {
+      const auto parse_result_right = _parse_predicate(base_node);
+
+      const auto and_predicate = std::make_shared<JoinPlanLogicalPredicate>(
+          left_predicate, JoinPlanPredicateLogicalOperator::And, parse_result_right.predicate);
+
+      return {parse_result_right.base_node, and_predicate};
+    }
+  } else if (node->type() == LQPNodeType::Union) {
+    return _parse_union(std::static_pointer_cast<UnionNode>(node));
+  } else {
+    Fail("Unexpected node type");
+    return {nullptr, nullptr};
+  }
+}
+
+JoinGraphBuilder::PredicateParseResult JoinGraphBuilder::_parse_union(
+    const std::shared_ptr<UnionNode>& union_node) const {
```

Comment:
We should probably test this for nested ors, don't you think?

Reply:
Test what exactly? You mean that nested ors are detected correctly?

Reply:
Reading that again, that is most probably what you meant ;)

Reply:
Yepp. "a=1 or a=2 or a=3" results in two unions, correct?

Reply:
most precisely.

Reply:
added test for nested OR and one for BETWEEN